python/pyarrow/_table.pyx

```diff
@@ -540,15 +559,56 @@ cdef class RecordBatch:
 cdef table_to_blockmanager(const shared_ptr[CTable]& table, int nthreads):
     cdef:
         PyObject* result_obj
-        CColumn* col
-        int i
+        Column col
+        CTable* ctable = table.get()
+        size_t row_count = ctable.num_rows()
+        CSchema* schema = ctable.schema().get()
+        shared_ptr[CTable] block_table = table
+        list index_arrays = []
+        size_t i
+        size_t total_columns = ctable.num_columns()
+        list names = [
+            frombytes(ctable.column(i).get().name())
+            for i in range(total_columns)
+        ]
+        dict metadata = box_metadata(schema.metadata().get()) or {}
+        tuple indices = ()
+        tuple indices_have_name = ()
+        size_t number_of_index_levels
+        size_t index_start
+
+    if b'indices' in metadata:
+        indices = tuple(metadata[b'indices'])
+        indices_have_name = tuple(metadata[b'has_name'])
+        assert len(indices) == len(indices_have_name)
+
+    number_of_index_levels = len(indices)
+    index_start = total_columns - number_of_index_levels
+
+    for i, has_name in zip(indices, indices_have_name):
+        col = Column()
+        col.init(ctable.column(i))
+        idx = _pandas().Index(
+            col.to_pandas().values,
+            name=names[i] if has_name else None
+        )
+        index_arrays.append(idx)
+        ctable.RemoveColumn(i, &block_table)
+        del names[i]
+
+    cdef list axes = [
+        names[:index_start],
+        _pandas().MultiIndex.from_arrays(
+            index_arrays
+        ) if index_arrays else _pandas().RangeIndex(row_count),
+    ]
```

Comment:
Same comment as above re: doing this in pure Python. It would also encourage adding appropriate public APIs to `pyarrow.Table`. We already have `Table.remove_column`, so it is probably better to use that if possible.